### PR TITLE
Drill samples refactoring

### PR DESCRIPTION
**Pull request by gui-co**
**Description of work.**

This PR is a refactoring of the DrILL model. The samples and parameters are now separate classes and communicate with the graphical part through their own MVP.

This PR has to be seen as a maintenance task of the DrILL interface. It was motivated by the increase of demand concerning new features and capacities of the interface, especially on the sample/parameter side (groups, master sample, parameter validity check, ...). View and model classes were starting to become "god objects", hard to maintain and control.

**To test:**

The DrILL interface should work the same way.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Interface/ui/drill/model/DrillParameter.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+
+import numpy
+
+from qtpy.QtCore import QObject, Signal
+
+from mantid.kernel import *
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillParameter.py:12]: Just an early comment, could you import only what you need? * imports are tolerated only for `simpleapi`, as it contains only functions.

File: scripts/Interface/ui/drill/model/DrillParameter.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                self._type = self.STRING_TYPE
+        elif (isinstance(mantidProperty, BoolPropertyWithValue)):
+            self._type = self.BOOL_TYPE
+        elif (isinstance(mantidProperty, FloatArrayProperty)):
+            self._type = self.FLOAT_ARRAY_TYPE
+        elif (isinstance(mantidProperty, IntArrayProperty)):
+            self._type = self.INT_ARRAY_TYPE
+        else:
+            self._type = self.STRING_TYPE
+
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillParameter.py:125]: It looks like StringArray is missing.

File: scripts/Interface/ui/drill/model/DrillParameter.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        if isinstance(value, numpy.ndarray):
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillParameter.py:126]: I think what is meant here is only 1D array. More than one dimensions should not be allowed.

File: scripts/Interface/ui/drill/model/DrillParameter.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    def setValue(self, value):
+        """
+        Set the parameter value. If a controller is available, the value will
+        be checked.
+
+        Args:
+            value (any): value
+        """
+        self._value = value
+        if self._controller is not None:
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillParameter.py:151]:
```suggestion
        if self._controller:
```

File: scripts/Interface/ui/drill/model/DrillParameter.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    def setValidationState(self, valid, msg=""):
+        """
+        Set the parameter validation state.
+
+        Args:
+            valid (bool): True if the parameter is valid
+            msg (str): optional error message
+        """
+        self._valid = valid
+        if not valid:
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillParameter.py:173]:
```suggestion
        self._validationErrorMsg = msg if not valid else None
```

File: scripts/Interface/ui/drill/model/DrillParameter.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            self._validationErrorMsg = msg
+        else:
+            self._validationErrorMsg = None
+        self.checked.emit()
+
+    def isValid(self):
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillParameter.py:179]: Not critical, but you can use the property and setter annotations to make the code more explicit.

File: scripts/Interface/ui/drill/model/DrillParameterController.py [hunk 7]
```diff
@@ -145,17 +76,18 @@ def run(self):
...
+                    defaultValue = self._alg.getProperty(pName).getDefault
+                    self._alg.setProperty(pName, pValue)
                     try:
-                        self._alg.setProperty(p.name, defaultValue)
+                        self._alg.setProperty(pName, defaultValue)
                     except:
                         pass # in case of mandatory parameter
-                    self._signals.okParam.emit(p)
+                    p.setValidationState(True)
                 except Exception as e:
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillParameterController.py:90]: Can you narrow down the Exception type?

File: scripts/Interface/ui/drill/model/DrillRundexIO.py [hunk 8]
```diff
@@ -47,7 +46,7 @@ def getFilename(self):
         """
         return self._filename
 
-    def load(self):
+    def load(self):  # noqa
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillRundexIO.py:49]: This should be split and noqa removed.

File: scripts/Interface/ui/drill/model/DrillSample.py [hunk 9]
```diff
@@ -5,87 +5,349 @@
...
+    def setGroup(self, group, index=None):
+        """
+        Set the group and index of the sample.
+
+        Args:
+            group (str): name of the group. None if the sample is not in a
+                         group.
+            index (int): index of the sample in its group
+        """
+        if not group:
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillSample.py:173]: If both if/else branches are non-empty, it is more readable to have the direct condition. Double negation causes a delay in reading the code.

File: scripts/Interface/ui/drill/model/DrillSample.py [hunk 10]
```diff
@@ -5,87 +5,349 @@
...
+        return True
+
+    def getParameterValues(self):
+        """
+        Get the parameters as a dictionnary.
+
+        Returns:
+            dict(str: str): parameters
+        """
+        out = dict()
```
Comment (gvardany) [inline scripts/Interface/ui/drill/model/DrillSample.py:301]:
```suggestion
        return dict( (name, param.getValue()) for name, param in self._parameters.items())
```

File: scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+
+from qtpy.QtGui import QBrush, QColor
+from qtpy.QtCore import *
```
Comment (gvardany) [inline scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py:9]: Import what you use.

File: scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
```
Comment (gvardany) [inline scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py:3]: You should put 2021 for the new files.

File: scripts/Interface/ui/drill/presenter/DrillPresenter.py [hunk 13]
```diff
@@ -513,73 +399,76 @@ def onClose(self):
...
+        Reset the table header.
+        """
+        acquisitionMode = self.model.getAcquisitionMode()
+        if acquisitionMode not in RundexSettings.COLUMNS:
+            self.view.set_table([], [])
+            return False
+        parameters = self.model.getParameters()
+        columns = RundexSettings.COLUMNS[acquisitionMode]
+        tooltips = list()
+        for name in columns:
```
Comment (gvardany) [inline scripts/Interface/ui/drill/presenter/DrillPresenter.py:468]: This looks redundant. Can one param match more than one column? if not break. Again, you can have a small function that returns the doc of a parameter by name.

**Review by gvardany — COMMENTED** [review]
Code seems more modular now, just few remarks. Will proceed to functional tests shortly.

File: scripts/Interface/ui/drill/model/DrillParameter.py (hunk 1 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/model/DrillParameter.py]: Done

File: scripts/Interface/ui/drill/model/DrillParameter.py (hunk 2 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/model/DrillParameter.py:130]: I added it

File: scripts/Interface/ui/drill/model/DrillParameter.py (hunk 3 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/model/DrillParameter.py:131]: I added a check for the dimension

File: scripts/Interface/ui/drill/model/DrillParameter.py (hunk 4 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/model/DrillParameter.py:157]: I followed this https://developer.mantidproject.org/Standards/PythonStandards.html#none-checks

File: scripts/Interface/ui/drill/model/DrillParameter.py (hunk 5 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/model/DrillParameter.py]: Done

File: scripts/Interface/ui/drill/model/DrillSample.py (hunk 10 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/model/DrillSample.py]: Done

File: scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py (hunk 11 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py]: Done

File: scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py (hunk 12 above)
Comment (gui-co) [reply scripts/Interface/ui/drill/presenter/DrillParameterPresenter.py]: Done

**Comment by gvardany** [comment]
The issues are fixed. The remaining problem is the parameter validation; if a rundex is loading while not having the correct data directories, it freezes until all the validation is completed. It is not really run asynchronously. I'll see what can be done about that.

**Comment by gvardany** [comment]
Apparently this was already the behaviour previously, so it's not a regression in this PR. The difference is that now the cell validation is launched also when loading a rundex, which wasn't the case before; something that made the issue more apparent.

**Comment by gvardany** [comment]
Investigating this further, the issue was that if one has many data search directories pointing to the network drive, except the right one where the data is, the file finding/validation itself is very slow (see https://github.com/mantidproject/mantid/issues/29881). In this case, the interface gets effectively blocked due to sequential validation of cells, each of which takes long. Considering that this is a rare scenario and that it can be averted by adding the data directories before loading the rundex, I think this PR shouldn't be blocked. Another improvement would be to make data search directories portable, which I added as an item to this issue, https://github.com/mantidproject/mantid/issues/31403.

**Review by gvardany — APPROVED** [review]
Agreed that the developer documentation will be added separately.

**Pull request merged**